Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -2528,18 +2528,28 @@ RefPtr<StyleValue> Parser::parse_string_value(ComponentValue const& component_va
     return nullptr;
 }
 
-RefPtr<StyleValue> Parser::parse_image_value(ComponentValue const& component_value)
+RefPtr<StyleValue> Parser::parse_image_value(TokenStream<ComponentValue>& tokens)
 {
-    auto url = parse_url_function(component_value);
-    if (url.has_value())
+    auto transaction = tokens.begin_transaction();
+    auto& token = tokens.next_token();
+
+    if (auto url = parse_url_function(token); url.has_value()) {
+        transaction.commit();
         return ImageStyleValue::create(url.value());
-    auto linear_gradient = parse_linear_gradient_function(component_value);
-    if (linear_gradient)
+    }
+    if (auto linear_gradient = parse_linear_gradient_function(token)) {
+        transaction.commit();
         return linear_gradient;
-    auto conic_gradient = parse_conic_gradient_function(component_value);
-    if (conic_gradient)
+    }
+    if (auto conic_gradient = parse_conic_gradient_function(token)) {
+        transaction.commit();
         return conic_gradient;
-    return parse_radial_gradient_function(component_value);
+    }
+    if (auto radial_gradient = parse_radial_gradient_function(token)) {
+        transaction.commit();
+        return radial_gradient;
+    }
+    return nullptr;
```

Comment:
Seeing this pattern how about some RAII helper like this
```c++
auto transaction = tokens.transaction();

// Happy Paths
...
...

// Abort
transaction.abort();
return nullptr;
```

Reply:
Seeing other functions with more complex paths, some might profit off of auto committing while others are nicer with the current auto roll-back/abort

Reply:
:thinking: Yeah that might be nice... Though this function in particular eventually won't need a transaction at all, because the functions it calls into will also take a TokenStream, and only progress it if they succeed. There's a fair amount of temporary messiness here until everything works the same way.

Reply:
OK, after looking through all the uses of `commit()`, there are a couple of situations like this one, but that's kind of all. It's verbose in those cases, but I think it'd be more confusing to have to deal with two kinds of transaction with opposite behaviours, and remember which type each one is. As things are, the pattern is generally to commit and then return something - and that makes it easy to spot places where we don't commit, because they're different.

I'm going to mark this as resolved, but thanks, it was well worth thinking about.

Also noticed I definitely could be smarter in places about how I use transactions, but that's not relevant to this. :sweat_smile: